packages/datadog-instrumentations/src/mocha/main.js

```diff
@@ -0,0 +1,481 @@
+
+      isSuitesSkipped = suitesToRun.length !== runner.suite.suites.length
+
+      log.debug(
+        () => `${suitesToRun.length} out of ${runner.suite.suites.length} suites are going to run.`
+      )
+
+      runner.suite.suites = suitesToRun
+
+      skippedSuites = Array.from(filteredSuites.skippedSuites)
+
+      global.run()
+    }
+
+    const onReceivedKnownTests = ({ err, knownTests: receivedKnownTests }) => {
+      if (err) {
+        knownTests = []
+        isEarlyFlakeDetectionEnabled = false
+      } else {
+        knownTests = receivedKnownTests
+      }
+
+      if (isSuitesSkippingEnabled) {
+        skippableSuitesCh.publish({
+          onDone: mochaRunAsyncResource.bind(onReceivedSkippableSuites)
+        })
+      } else {
+        global.run()
+      }
+    }
+
+    const onReceivedConfiguration = ({ err, libraryConfig }) => {
+      if (err || !skippableSuitesCh.hasSubscribers || !knownTestsCh.hasSubscribers) {
+        return global.run()
+      }
+
+      isEarlyFlakeDetectionEnabled = libraryConfig.isEarlyFlakeDetectionEnabled
+      isSuitesSkippingEnabled = libraryConfig.isSuitesSkippingEnabled
+      earlyFlakeDetectionNumRetries = libraryConfig.earlyFlakeDetectionNumRetries
+
+      if (isEarlyFlakeDetectionEnabled) {
+        knownTestsCh.publish({
+          onDone: mochaRunAsyncResource.bind(onReceivedKnownTests)
+        })
+      } else if (isSuitesSkippingEnabled) {
+        skippableSuitesCh.publish({
+          onDone: mochaRunAsyncResource.bind(onReceivedSkippableSuites)
+        })
+      } else {
+        global.run()
+      }
+    }
+
+    mochaRunAsyncResource.runInAsyncScope(() => {
+      libraryConfigurationCh.publish({
+        onDone: mochaRunAsyncResource.bind(onReceivedConfiguration)
+      })
+    })
+
+    return runner
+  })
+  return Mocha
+})
+
+// Only used to set `mocha.options.delay` to true in serial mode. When the mocha CLI is used,
+// setting options.delay in Mocha#run is not enough to delay the execution.
+// TODO: modify this hook to grab the data in parallel mode, so that ITR and EFD can work.
+addHook({
+  name: 'mocha',
+  versions: ['>=5.2.0'],
+  file: 'lib/cli/run-helpers.js'
+}, (run) => {
+  shimmer.wrap(run, 'runMocha', runMocha => async function () {
+    if (!testStartCh.hasSubscribers) {
+      return runMocha.apply(this, arguments)
+    }
+    const mocha = arguments[0]
+    /**
+     * This attaches `run` to the global context, which we'll call after
+     * our configuration and skippable suites requests
+     */
+    if (!mocha.options.parallel) {
+      mocha.options.delay = true
+    }
+    return runMocha.apply(this, arguments)
+  })
+  return run
+})
+
+// Only used in serial mode (no --parallel flag is passed)
+// This hook is used to generate session, module, suite and test events
+addHook({
+  name: 'mocha',
+  versions: ['>=5.2.0'],
+  file: 'lib/runner.js'
+}, function (Runner, frameworkVersion) {
+  if (patched.has(Runner)) return Runner
```

Comment:
ℹ️ the logic on this hook was not modified, but we extracted some of the mocha event handlers to functions in `mocha/utils.js`: `getOnStartHandler` etc.